docusaurus/docs/how-to-guides/data-source-plugins/add-migration-handler-for-backend.md

```diff
@@ -0,0 +1,216 @@
++                       return backend.DataResponse{}, err
+                }
+                q := req.URL.Query()
+                q.Add("multiplier", strconv.Itoa(input.Multiply))
+...
+
++func convertQuery(orig backend.DataQuery) (*kinds.DataQuery, error) {
++       input := &kinds.DataQuery{}
++       err := json.Unmarshal(orig.JSON, input)
++       if err != nil {
++               return nil, fmt.Errorf("unmarshal: %w", err)
++       }
++       if input.Multiplier != 0 && input.Multiply == 0 {
++               input.Multiply = input.Multiplier
++               input.Multiplier = 0
++       }
++       return input, nil
++}
+```
+
+### Step 5: Use migration code from the frontend (using experimental APIs)
+
+:::note
+
+This feature depends on the feature flag [grafanaAPIServerWithExperimentalAPIs](https://github.com/grafana/grafana/blob/build-go-fast/pkg/services/featuremgmt/registry.go#L528). It also requires the package **@grafana/runtime > 11.4** (still experimental functionality). If your plugin implements this feature, bump its **grafanaDepencency to ">=11.4.0"**. If your plugin can't adhere to these requirements, refer to [Run migrations using legacy APIs](#run-migrations-using-legacy-apis).
+
+:::
+
+You should be able to invoke your `convertQuery` function from the frontend as well as the backend, so our `QueryEditor` component should be able to convert the query to the new format. In order to expose this function to the frontend, the backend needs to implement the `QueryConversionHandler` interface. This is just a wrapper around the `convertQuery` function, but for multiple queries. Refer to how this is implemented in [this example](https://github.com/grafana/grafana-plugin-examples/pull/403/files#diff-a4ecf4de6bfd217cb4e941334f2b203adb8b400b9784c8fa94f17847f7570871).
+
+Finally, adapt the frontend so `@grafana/runtime` knows if it should run the migration action. Do this in two steps:
+
+1. Implement the `MigrationHandler` that `@grafana/runtime` exposes in the plugin's `DataSource` class. Set the property `hasBackendMigration` (to `true`) and implement the function `shouldMigrate`. The `shouldMigrate` function receives a query and verifies if it requires migration (for example, by checking the latest properties or checking the expected plugin version, if it's part of the model). This verification avoids unnecessary queries to the backend.
+1. Use the wrapper `QueryEditorWithMigration` along with your `QueryEditor` component. This wrapper will ensure that the query is migrated before rendering the editor.
+
+That's it. Once the plugin implements these steps, existing and new queries will continue working without the need for duplicate migration logic in multiple places.
+
+:::note
+
+To see how steps 2 to 5 are done in a complete example, refer to [this example](https://github.com/grafana/grafana-plugin-examples/pull/403/files).
+
+:::
+
+#### Run migrations using legacy APIs
```

Comment:
I am not sure that this section is marked clearly that it is an alternative not an additional step to run in Step 5.

Reply:
I agree (more after seeing the preview). Before I had this at the same level than `Step 5`. I had this as `Step 5'`, which is maybe a confusing notation 😅 . @josmperez suggested this. Let me propose `Step 5 (alternative): ...`. I think that will be more clear.

Reply:
Step 5 (alternative) is good. unfortunately h4s look only link regular bold text in our style sheet. we should really redefine styles for h4 to make them pop out.